README.md
[![Build Status](https://travis-ci.org/qubole/kinesis-sql.svg?branch=master)](https://travis-ci.org/qubole/kinesis-sql)

## NOTE: This project is NO LONGER MAINTAINED. 

[Ron Cremer](https://github.com/roncemer) has volunteered to maintain this project. Beginning with Spark 3.2, the new project is located here: https://github.com/roncemer/spark-sql-kinesis


# Kinesis Connector for Structured Streaming 

Implementation of Kinesis Source Provider in Spark Structured Streaming. [SPARK-18165](https://issues.apache.org/jira/browse/SPARK-18165) describes the need for such implementation. More details on the implementation can be read in this [blog](https://www.qubole.com/blog/kinesis-connector-for-structured-streaming/)

## Downloading and Using the Connector

The connector is available from the Maven Central repository. It can be used using the --packages option or the spark.jars.packages configuration property. Use the following connector artifact

	Spark 3.0: com.qubole.spark/spark-sql-kinesis_2.12/1.2.0-spark_3.0
	Spark 2.4: com.qubole.spark/spark-sql-kinesis_2.11/1.2.0-spark_2.4

## Developer Setup
Checkout kinesis-sql branch depending upon your Spark version. Use Master branch for the latest Spark version 

###### Spark version 3.0.x
	git clone git@github.com:qubole/kinesis-sql.git
	git checkout master
	cd kinesis-sql
	mvn install -DskipTests

This will create *target/spark-sql-kinesis_2.12-\*.jar* file which contains the connector code and its dependency jars.


## How to use it

#### Setup Kinesis
Refer [Amazon Docs](https://docs.aws.amazon.com/cli/latest/reference/kinesis/create-stream.html) for more options

###### Create Kinesis Stream 

	$ aws kinesis create-stream --stream-name test --shard-count 2
    
###### Add Records in the stream
	
    $ aws kinesis put-record --stream-name test --partition-key 1 --data 'Kinesis'
    $ aws kinesis put-record --stream-name test --partition-key 1 --data 'Connector'
    $ aws kinesis put-record --stream-name test --partition-key 1 --data 'for'
    $ aws kinesis put-record --stream-name test --partition-key 1 --data 'Apache'
    $ aws kinesis put-record --stream-name test --partition-key 1 --data 'Spark'

#### Example Streaming Job

Refering $SPARK_HOME to the Spark installation directory.

###### Open Spark-Shell

	$SPARK_HOME/bin/spark-shell --jars target/spark-sql-kinesis_2.11-2.2.0.jar

###### Subscribe to Kinesis Source
	// Subscribe the "test" stream
	scala> :paste
	val kinesis = spark
  		.readStream
  		.format("kinesis")
    	.option("streamName", "spark-streaming-example")
       	.option("endpointUrl", "https://kinesis.us-east-1.amazonaws.com")
        .option("awsAccessKeyId", [ACCESS_KEY])
        .option("awsSecretKey", [SECRET_KEY])
        .option("startingposition", "TRIM_HORIZON")
  		.load

###### Check Schema 
	scala> kinesis.printSchema
	root
 	|-- data: binary (nullable = true)
 	|-- streamName: string (nullable = true)
 	|-- partitionKey: string (nullable = true)
 	|-- sequenceNumber: string (nullable = true)
 	|-- approximateArrivalTimestamp: timestamp (nullable = true)

###### Word Count 
	// Cast data into string and group by data column
	scala> :paste
    
    	 kinesis
        .selectExpr("CAST(data AS STRING)").as[(String)]
        .groupBy("data").count()
  		.writeStream
  		.format("console")
        .outputMode("complete") 
  		.start()
  		.awaitTermination()
        
###### Output in Console


	+------------+-----+
	|        data|count|
	+------------+-----+
	|         for|    1|
	|      Apache|    1|
    |       Spark|    1|
	|     Kinesis|    1|
	|   Connector|    1|
	+------------+-----+ 

###### Using the Kinesis Sink
    // Cast data into string and group by data column
        scala> :paste
        kinesis
        .selectExpr("CAST(rand() AS STRING) as partitionKey","CAST(data AS STRING)").as[(String,String)]
        .groupBy("data").count()
  	    .writeStream
  	    .format("kinesis")
        .outputMode("update") 
        .option("streamName", "spark-sink-example")
        .option("endpointUrl", "https://kinesis.us-east-1.amazonaws.com")
        .option("awsAccessKeyId", [ACCESS_KEY])
        .option("awsSecretKey", [SECRET_KEY])
  	    .start()
  	    .awaitTermination()

## Kinesis Source Configuration 

 Option-Name        | Default-Value           | Description  |
| ------------- |:-------------:| -----:|
| streamName     | - | Name of the stream in Kinesis to read from |
| endpointUrl     |   https://kinesis.us-east-1.amazonaws.com    |   end-point URL for Kinesis Stream|
| awsAccessKeyId |    -     |    AWS Credentials for Kinesis describe, read record operations |
| awsSecretKey |      -  |    AWS Credentials for Kinesis describe, read record operations |
| awsSTSRoleARN |      -  |    AWS STS Role ARN for Kinesis describe, read record operations |
| awsSTSSessionName |      -  |    AWS STS Session name for Kinesis describe, read record operations |
| awsUseInstanceProfile | true |    Use Instance Profile Credentials if none of credentials provided |
| startingPosition |      LATEST |    Starting Position in Kinesis to fetch data from. Possible values are "latest", "trim_horizon", "earliest" (alias for trim_horizon), or JSON serialized map shardId->KinesisPosition   |
| failondataloss| true | fail the streaming job if any active shard is missing or expired
| kinesis.executor.maxFetchTimeInMs |     1000 |  Maximum time spent in executor to fetch record from Kinesis per Shard |
| kinesis.executor.maxFetchRecordsPerShard |     100000 |  Maximum Number of records to fetch per shard  |
| kinesis.executor.maxRecordPerRead |     10000 |  Maximum Number of records to fetch per getRecords API call  |
| kinesis.executor.addIdleTimeBetweenReads	| false	| Add delay between two consecutive getRecords API call	|
| kinesis.executor.idleTimeBetweenReadsInMs	| 1000	| Minimum delay between two consecutive getRecords	| 
| kinesis.client.describeShardInterval |      1s (1 second) |  Minimum Interval between two ListShards API calls to consider resharding  |
| kinesis.client.numRetries |     3 |  Maximum Number of retries for Kinesis API requests  |
| kinesis.client.retryIntervalMs |     1000 |  Cool-off period before retrying Kinesis API  |
| kinesis.client.maxRetryIntervalMs	| 10000	| Max Cool-off period between 2 retries	|
| kinesis.client.avoidEmptyBatches| false | Avoid creating an empty microbatch job by checking upfront if there are any unread data in the stream before the batch is started

## Kinesis Sink Configuration
 Option-Name        | Default-Value           | Description  |
| ------------- |:-------------:| -----:|
| streamName   | - | Name of the stream in Kinesis to write to|
| endpointUrl  | https://kinesis.us-east-1.amazonaws.com |  The aws endpoint of the kinesis Stream |
| awsAccessKeyId |    -     |    AWS Credentials for  Kinesis describe, read record operations    
| awsSecretKey |      -  |    AWS Credentials for  Kinesis describe, read record |
| awsSTSRoleARN |      -  |    AWS STS Role ARN for Kinesis describe, read record operations |
| awsSTSSessionName |      -  |    AWS STS Session name for Kinesis describe, read record operations |
| awsUseInstanceProfile | true |    Use Instance Profile Credentials if none of credentials provided |
| kinesis.executor.recordMaxBufferedTime | 1000 (millis) | Specify the maximum buffered time of a record |
| kinesis.executor.maxConnections | 1 | Specify the maximum connections to Kinesis | 
| kinesis.executor.aggregationEnabled | true | Specify if records should be aggregated before sending them to Kinesis |
| kniesis.executor.flushwaittimemillis | 100 | Wait time while flushing records to Kinesis on Task End |

## Roadmap
*  We need to migrate to DataSource V2 APIs for MicroBatchExecution.
*  Maintain Per Micro-Batch Shard Commit state in Dynamo DB

## Acknowledgement

This connector would not have been possible without reference implemetation of [Kafka connector](https://github.com/apache/spark/tree/branch-2.2/external/kafka-0-10-sql) for Structured streaming, [Kinesis Connector](https://github.com/apache/spark/tree/branch-2.2/external/kinesis-asl) for Legacy Streaming and [Kinesis Client Library](https://github.com/awslabs/amazon-kinesis-client). Structure of some part of the code is influenced by the excellent work done by various Apache Spark Contributors.


CHANGELOG
Changelog
=========

spark-sql-kinesis_2.13-1.2.1_spark-3.2
----------------------------
- Update README.md to indicate that this repo is no longer maintained, and include a link to the new repo.

kinesis_2.12-1.2.0_spark-3.0
----------------------------
- Support for Spark 3.0 and scala 2.12 (#92) [Vikram Agrawal, Vikram
  Agrawal]
- Switch to use Kinesis list-shards API (#89) [Chad Lagore]
- Fix issue with slow kinesis sink. [Vikram Agrawal]
- Remove lastReadSequenceNumber.isEmpty condition (#90) [Chad Lagore]
- Fixing latency in kinesis sink (#81) [abhishekd0907]

kinesis_2.11-1.1.4-spark_2.4
----------------------------
- Fix retries while making Kinesis API calls (#79) [Vikram Agrawal]
- Adding ability to set exact Kinesis position to start reading from
  (#78) [Stanislav Norochevskyi]

  * Adding ability read initial Kinesis position from checkpoint JSON representation.

  * Adding a unit test for InitialKinesisPosition parser
- Apply retries to kinesis exceptions with status code >= 500 (#76)
  [Arne Huang]
- Make DefaultAWSCredentialsProviderChain available to choose (#71)
  [Hyeonseop Lee]

  * Add option to use instance profile credentials

  * Update docs

  * Add same option to producer
- Added session based authentication (#72) [guoming-xu]
- Add source and target java version to overwrite defaults; add nobootcp
  option to scala compiler (#68) [Julian Keppel]
- Update the artifact of the new release. [Vikram Agrawal]
- [maven-release-plugin] prepare for next development iteration. [Vikram
  Agrawal]
- Merge branch 'master' of github.com:qubole/kinesis-sql. [Vikram
  Agrawal]
- Update README.md. [Vikram Agrawal]
- Update README.md. [Vikram Agrawal]
- Update README.md. [Vikram Agrawal]

kinesis_2.11-1.1.3-spark_2.4
--------------------------------
- [maven-release-plugin] prepare release spark-sql-
  kinesis_2.11-1.1.3-spark_2.4. [Vikram Agrawal]
- GitChangelog RC file and changelog. [Vikram Agrawal]
- Add travis build config. [Vikram Agrawal]
- Create fat jar with dependencies shaded in the jar. [Vikram Agrawal]
- Handle deleted shards (#59) [Vikram Agrawal]
- [maven-release-plugin] prepare for next development iteration. [Vikram
  Agrawal]

kinesis_2.11-1.1.2-spark_2.4
-----------------------------
- [maven-release-plugin] prepare release spark-sql-
  kinesis_2.11-1.1.2-spark_2.4. [Vikram Agrawal]
- [maven-release-plugin] prepare for next development iteration. [Vikram
  Agrawal]
- POM changes. [Vikram Agrawal]
- Update pom.xml. [Vikram Agrawal]
- [maven-release-plugin] prepare for next development iteration. [Vikram
  Agrawal]
- [maven-release-plugin] prepare release spark-sql-
  kinesis_2.11-1.1.0-spark_2.4. [Vikram Agrawal]
- Update Pom file for mvn release. [Vikram Agrawal]
- Fix Deserialisation/Serialisation of KinesisOffsets  (#62) [Vikram
  Agrawal]

  * Fix incorrect metadata for batchId in the kinesis offsets

  * Fix serialization and deserialization fo the kinesis offsets
- Fix issue with Reprocessing oldshards issue (#63) [Vikram Agrawal]
- 2.4.0 (#56) [Vikram Agrawal]

  * Fix broken test suite and scalastyle error

  *  Fix issues with over-ridden Kinesis Source Options (#36)

  * Fix issues with over-ridden Kinesis Source Options

  * Update ReadMe

  * Fix Review Comments

  * Fix ambiguity with scala 2.12 (#37)

  * Fix stylecheck errors

  * Add retries to Handle EC issues in HDFS metadata committer (#41)

  * Handle EC issues in HDFS MetaCommitter

  * Avoid timestamp as Offset and avoid empty batches when there is no new data (#49)
- Changing protobuf version and fixing styling errors (#53)
  [abhishekd0907]
- 2.4.0 (#45) [nhampiholi]

  * Fix broken test suite and scalastyle error

  *  Fix issues with over-ridden Kinesis Source Options (#36)

  * Fix issues with over-ridden Kinesis Source Options

  * Update ReadMe

  * Fix Review Comments

  * Fix ambiguity with scala 2.12 (#37)

  * Fix stylecheck errors

  * Add retries to Handle EC issues in HDFS metadata committer (#41)

  * Handle EC issues in HDFS MetaCommitter

  * Avoid timestamp as Offset and avoid empty batches when there is no new data (#49)
- Avoid timestamp as Offset and avoid empty batches when there is no new
  data (#49) [Vikram Agrawal]
- Add retries to Handle EC issues in HDFS metadata committer (#41)
  [Vikram Agrawal]

  * Handle EC issues in HDFS MetaCommitter
- Fix stylecheck errors. [Vikram Agrawal]
- Fix ambiguity with scala 2.12 (#37) [Marcin Szymański]
- Fix issues with over-ridden Kinesis Source Options (#36) [Vikram
  Agrawal]

  * Fix issues with over-ridden Kinesis Source Options

  * Update ReadMe

  * Fix Review Comments
- Fix broken test suite and scalastyle error. [anthony.may]
- Use latest DSv2 APIs for continuous processing. [Vikram Agrawal]
- Handle greater than 100 shard streams. [Vikram Agrawal]
- Fix for handling greater than 100 shard streams. [Siddhartha Jain]
- Update README.md. [Vikram Agrawal]
- Add AWSInstanceProfileCredentialsProviderWithRetries to handle issues
  with instance profile provider. [Vikram Agrawal]
- Bump up spark version. [Vikram Agrawal]
- Support of Kinesis Connector for Continuous streaming (#15) [Vikram
  Agrawal]
- Fixed aws sdk and kpl versions (#18) [Georgios]
- Bump up Kinesis Client version. [Vikram Agrawal]
- Support of AWS roles / instance profile for Authentication (#13)
  [Vikram Agrawal]

  * Support of AWS roles / instance profile for Authentication

  * InstanceProfileCredentials should be case object instead of case class

  * Fix unit tests

  * clean up
- Update README. [Vikram Agrawal]
- Merge branch '2.3.0' [Vikram Agrawal]
- Fixes to support build against SPARK 2.3.0. [Vikram Agrawal]
- Merge pull request #8 from ggeorgiadis/master. [Vikram Agrawal]

  Fixed issue with closed child process caused by the cache timeout
- Cleanup. [Georgios Georgiadis]
- Fixed bug in parameters validation and cache timeout. [Georgios
  Georgiadis]
- Merge pull request #5 from ggeorgiadis/master. [Vikram Agrawal]


src/main/scala/org/apache/spark/sql/kinesis/KinesisReader.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.spark.sql.kinesis

import java.math.BigInteger
import java.util
import java.util.{ArrayList, Locale}
import java.util.concurrent.{Executors, ThreadFactory}

import com.amazonaws.{AbortedException, ClientConfiguration, ClientConfigurationFactory}
import com.amazonaws.services.kinesis.AmazonKinesisClient
import com.amazonaws.services.kinesis.clientlibrary.types.UserRecord
import com.amazonaws.services.kinesis.model.{GetRecordsRequest, ListShardsRequest, Shard, _}
import scala.collection.JavaConverters._
import scala.concurrent.{ExecutionContext, Future}
import scala.concurrent.duration.Duration
import scala.util.control.NonFatal

import org.apache.spark.internal.Logging
import org.apache.spark.sql.types._
import org.apache.spark.util.{ThreadUtils, UninterruptibleThread}


// This class uses Kinesis API to read data offsets from Kinesis

private[kinesis] case class KinesisReader(
    readerOptions: Map[String, String],
    streamName: String,
    kinesisCredsProvider: SparkAWSCredentials,
    endpointUrl: String
) extends Serializable with Logging {

  /*
   * Used to ensure execute fetch operations execute in an UninterruptibleThread
   */
  val kinesisReaderThread = Executors.newSingleThreadExecutor(new ThreadFactory {
    override def newThread(r: Runnable): Thread = {
      val t = new UninterruptibleThread("Kinesis Reader") {
        override def run(): Unit = {
          r.run()
        }
      }
      t.setDaemon(true)
      t
    }
  })

  val execContext = ExecutionContext.fromExecutorService(kinesisReaderThread)

  private val maxOffsetFetchAttempts =
    readerOptions.getOrElse("client.numRetries".toLowerCase(Locale.ROOT), "3").toInt

  private val offsetFetchAttemptIntervalMs =
    readerOptions.getOrElse("client.retryIntervalMs".toLowerCase(Locale.ROOT), "1000").toLong

  private val maxRetryIntervalMs: Long = {
    readerOptions.getOrElse("client.maxRetryIntervalMs".toLowerCase(Locale.ROOT), "10000").toLong
  }

  private val clientExecutionTimeout: Int = {
    readerOptions.get("client.clientExecutionTimeout".toLowerCase(Locale.ROOT))
      .fold(ClientConfiguration.DEFAULT_CLIENT_EXECUTION_TIMEOUT)(_.toInt)
  }

  private val maxSupportedShardsPerStream = 10000;

  private var _amazonClient: AmazonKinesisClient = null

  private def getAmazonClient(): AmazonKinesisClient = {
    if (_amazonClient == null) {
      val config = new ClientConfigurationFactory().getConfig
      config.setClientExecutionTimeout(clientExecutionTimeout)
      _amazonClient = new AmazonKinesisClient(kinesisCredsProvider.provider, config)
      _amazonClient.setEndpoint(endpointUrl)
    }
    _amazonClient
  }

  def getShards(): Seq[Shard] = {
    val shards = listShards
    logInfo(s"List shards in Kinesis Stream:  ${shards}")
    shards
  }

  def close(): Unit = {
    runUninterruptibly {
      if (_amazonClient != null) {
        _amazonClient.shutdown()
        _amazonClient = null
      }
    }
    kinesisReaderThread.shutdown()
  }

  def getShardIterator(shardId: String,
                       iteratorType: String,
                       iteratorPosition: String,
                       failOnDataLoss: Boolean = true): String = {

    val getShardIteratorRequest = new GetShardIteratorRequest
    getShardIteratorRequest.setShardId(shardId)
    getShardIteratorRequest.setStreamName(streamName)
    getShardIteratorRequest.setShardIteratorType(iteratorType)

    if (iteratorType == "AFTER_SEQUENCE_NUMBER" || iteratorType == "AT_SEQUENCE_NUMBER") {
      getShardIteratorRequest.setStartingSequenceNumber(iteratorPosition)
    }

    if (iteratorType == "AT_TIMESTAMP") {
      logDebug(s"TimeStamp while getting shard iterator ${
        (new java.util.Date(iteratorPosition.toLong)).toString}")
      getShardIteratorRequest.setTimestamp(new java.util.Date(iteratorPosition.toLong))
    }

    runUninterruptibly {
      retryOrTimeout[GetShardIteratorResult](
        s"Fetching Shard Iterator") {
        try {
          getAmazonClient.getShardIterator(getShardIteratorRequest)
        } catch {
          case r: ResourceNotFoundException =>
            if (!failOnDataLoss) {
              new GetShardIteratorResult()
            }
            else {
              throw r
            }
        }
      }
    }.getShardIterator
  }


  def getKinesisRecords(shardIterator: String, limit: Int): GetRecordsResult = {
    val getRecordsRequest = new GetRecordsRequest
    getRecordsRequest.setShardIterator(shardIterator)
    getRecordsRequest.setLimit(limit)
    val getRecordsResult: GetRecordsResult = runUninterruptibly {
      retryOrTimeout[ GetRecordsResult ](s"get Records for a shard ") {
        getAmazonClient.getRecords(getRecordsRequest)
      }
    }
    getRecordsResult
  }


  def deaggregateRecords(records: util.List[ Record ], shard: Shard): util.List[ Record] = {
    // We deaggregate if and only if we got actual Kinesis records, i.e.
    // not instances of some subclass thereof.
    if ( !records.isEmpty && records.get(0).getClass.equals(classOf[ Record ]) ) {
      if ( shard != null ) {
        return UserRecord.deaggregate(
          records,
          new BigInteger(shard.getHashKeyRange.getStartingHashKey),
          new BigInteger(shard.getHashKeyRange.getEndingHashKey))
          .asInstanceOf[ util.List[ _ ] ].asInstanceOf[ util.List[ Record ] ]
      } else {
        return UserRecord.deaggregate(records)
          .asInstanceOf[ util.List[ _ ] ].asInstanceOf[ util.List[ Record ] ]
      }
    }
    records
  }

  private def listShards(): Seq[Shard] = {
    var nextToken = ""
    var returnedToken = ""
    val shards = new ArrayList[Shard]()
    val listShardsRequest = new ListShardsRequest
    listShardsRequest.setStreamName(streamName)
    listShardsRequest.setMaxResults(maxSupportedShardsPerStream)

    do {
      val listShardsResult: ListShardsResult = runUninterruptibly {
        retryOrTimeout[ListShardsResult]( s"List shards") {
            getAmazonClient.listShards(listShardsRequest)
        }
      }
      shards.addAll(listShardsResult.getShards)
      returnedToken = listShardsResult.getNextToken()
      if (returnedToken != null) {
        nextToken = returnedToken
        listShardsRequest.setNextToken(nextToken)
      }
    } while (!nextToken.isEmpty)

    shards.asScala.toSeq
  }

  /*
   * This method ensures that the closure is called in an [[UninterruptibleThread]].
   * This is required when communicating with the AWS. In the case
   */
  private def runUninterruptibly[T](body: => T): T = {
    if (!Thread.currentThread.isInstanceOf[UninterruptibleThread]) {
      val future = Future {
        body
      }(execContext)
      ThreadUtils.awaitResult(future, Duration.Inf)
    } else {
      body
    }
  }

  /** Helper method to retry Kinesis API request with exponential backoff and timeouts */
  private def retryOrTimeout[T](message: String)(body: => T): T = {
    assert(Thread.currentThread().isInstanceOf[UninterruptibleThread])

    val startTimeMs = System.currentTimeMillis()
    var retryCount = 0
    var result: Option[T] = None
    var lastError: Throwable = null
    var waitTimeInterval = offsetFetchAttemptIntervalMs

    def isMaxRetryDone = retryCount >= maxOffsetFetchAttempts

    while (result.isEmpty && !isMaxRetryDone) {
      if ( retryCount > 0 ) { // wait only if this is a retry
        Thread.sleep(waitTimeInterval)
        waitTimeInterval = scala.math.min(waitTimeInterval * 2, maxRetryIntervalMs)
      }
      try {
        result = Some(body)
      } catch {
        case NonFatal(t) =>
          lastError = t
          t match {
            case ptee: ProvisionedThroughputExceededException =>
              logWarning(s"Error while $message [attempt = ${retryCount + 1}]", ptee)
            case lee: LimitExceededException =>
              logWarning(s"Error while $message [attempt = ${retryCount + 1}]", lee)
            case ae: AbortedException =>
              logWarning(s"Error while $message [attempt = ${retryCount + 1}]", ae)
            case ake: AmazonKinesisException =>
              if (ake.getStatusCode() >= 500) {
                logWarning(s"Error while $message [attempt = ${retryCount + 1}]", ake)
              } else {
                throw new IllegalStateException(s"Error while $message", ake)
              }
            case e: Throwable =>
              throw new IllegalStateException(s"Error while $message", e)
          }
      }
      retryCount += 1
    }
    result.getOrElse {
      throw new IllegalStateException(
        s"Gave up after $retryCount retries while $message, last exception: ", lastError)
    }
  }

}


private [kinesis]  object KinesisReader {

  val kinesisSchema: StructType =
      StructType(Seq(
        StructField("data", BinaryType),
        StructField("streamName", StringType),
        StructField("partitionKey", StringType),
        StructField("sequenceNumber", StringType),
        StructField("approximateArrivalTimestamp", TimestampType))
      )
}
